sqlite.go

```diff
@@ -172,26 +182,41 @@ type conn struct {
 	stmts    map[string]*stmt // persisted statements
 	txState  txState
 	readOnly bool
+	closed   atomic.Bool
 }
 
 func (c *conn) Prepare(query string) (driver.Stmt, error) { panic("deprecated, unused") }
 func (c *conn) Begin() (driver.Tx, error)                 { panic("deprecated, unused") }
 func (c *conn) Close() error {
+	// Don't double-close
+	if !c.closed.CompareAndSwap(false, true) {
+		UsesAfterClose.Add("Close", 1)
+		return nil
+	}
+
 	for q, s := range c.stmts {
 		s.stmt.Finalize()
+		s.closed.Store(true)
 		delete(c.stmts, q)
 	}
-	return reserr(c.db, "Conn.Close", "", c.db.Close())
+	err := reserr(c.db, "Conn.Close", "", c.db.Close())
+	return err
 }
 func (c *conn) PrepareContext(ctx context.Context, query string) (driver.Stmt, error) {
 	persist := ctx.Value(persistQuery{}) != nil
 	return c.prepare(ctx, query, persist)
 }
 
 func (c *conn) prepare(ctx context.Context, query string, persist bool) (s *stmt, err error) {
+	if c.closed.Load() {
+		UsesAfterClose.Add("prepare", 1)
+		return nil, ErrClosed
+	}
+
 	query = strings.TrimSpace(query)
 	if s := c.stmts[query]; s != nil {
 		s.prepCtx = ctx
+		s.closed.Store(false)
```

Comment:
should this be a CAS for paranoia? Going from false to false would indicate a bug?

Reply:
actually, I don't see where this is s.closed.Store(true) on return to the stmts pool?

Reply:
We do it on `Close` here:
https://github.com/tailscale/sqlite/pull/71/files#diff-3101fd8b5ebefc9739f7efdafa4dd6f04c19db98fc4428623ee619a2773c8501R446